### ISSUE TITLE
camonitor of a string array PV raises a TypeError exception if the array contains one string

### ISSUE THREAD

**Issue by hanak** [pyepics/pyepics#269]
Exception:
```
File "...\site-packages\epics\__init__.py", line 157, in camonitor
  thispv.get()
File "...\site-packages\epics\pv.py", line 486, in get
  data = self.get_with_metadata(count=count, as_string=as_string,
File "...\site-packages\epics\pv.py", line 48, in wrapped
  return func(self, *args, **kwargs)
File "...\site-packages\epics\pv.py", line 621, in get_with_metadata
  if count < len(val):
TypeError: len() of unsized object
```

The behavior is a result of interaction between following functions:
* `ca.unwap()` uses `ca.scan_string` to read a string. This function will return the string if retrieved list of strings contains just one string.
* `PV.get_with_metadata(as_numpy=True)` (the default used by `camonitor()`) does following:
  * Reads the string into a variable `val`
  * It tests that `val` supports `__len__`. In case of a string, this passes.
  * Then, it verifies `as_numpy` and `ca.HAS_NUMPY`. In this is satisfied. So, it converts `val` to a numpy object.
  * Then, it tries to run `len(val)`. Yet, `val` is a scalar at this moment (`val.shape` returns `()`). This raises an exception.

Further details:
* This has been observed on pyepics 3.4.3 using for automated tests. Yet, the code involved in this mechanism is the same at the current master branch.
* Standalone recreation of the scenario has not been performed because the PV is a custom record which mimics 
waveform (FTVL=STRING): it provides an array of strings. caget works with no complains.
* Use of `as_numpy=False` with `caget()` will return the first letter of the first string (i.e. `"R"` instead of `"READY"`) for an array containing `["READY"]`.

**Comment by newville** [comment]
@hanak Yeah, I can believe that.   We do not test with a custom record "which mimics waveform of strings".    

Can you give a more complete example of that record?   Like, is this an array of strings but there is actually only one element in the array? 

I'm sure we could figure out how to handle such cases. Ideally, we would want to add that to our tests.  Does this record use custom code or "just records"?

**Comment by hanak** [comment]
@newville Thank you for your response. Let me add some details:
* The custom record type provides the same fields as waveform (VAL, NELM, NORD) and it uses almost the same routines `cvt_dbaddr()` and `get_array_info()` as base waveform and EPICS tools caget/caput works with it. 
* NELM is 32, NORD is 1, and the record contains `["ITEM 1"]`. 

I was able to trigger the problem by using a plain waveform (see test `test_pv_unittests.py/test_get_string_waveform` in https://github.com/hanak/pyepics/tree/waveform-string-error). This test fails with:
```
FAILED tests/test_pv_unittests.py::test_get_string_waveform - TypeError: len() of unsized object
```

Can you confirm you see the same?

One I remove the case 1 from the pool of lengths, the test succeeded. Therefore, it seems to me this issue is generic rather than specific to the custom record type of mine. I think that the problem is a result of using `len()` as a tool for detecting an array combined with converting an array of one string to the string. My guess is the solution is to modify use of `len()`. Still, I am not able to tell whether this will not trigger a surprise somewhere else.

Environment:
* OS: Windows 10
* python 3.10.11
* pytest 8.2.2
* softIoc of EPICS base 7.0.6.1

**Comment by newville** [comment]
@newville Thank you for your response. Let me add some details:

> The custom record type provides the same fields as waveform (VAL, NELM, NORD) and it uses almost the same routines 
> cvt_dbaddr() and get_array_info() as base waveform and EPICS tools caget/caput works with it.
> NELM is 32, NORD is 1, and the record contains ["ITEM 1"].

Sorry, that is not enough to go on. If you would like us to support using such a record, you will need to post code
to allow someone else to create such a record. 

Is there such a thing as a waveform of strings?   


>  I was able to trigger the problem by using a plain waveform (see test test_pv_unittests.py/test_get_string_waveform in  
> https://github.com/hanak/pyepics/tree/waveform-string-error). This test fails with:
>
> FAILED tests/test_pv_unittests.py::test_get_string_waveform - TypeError: len() of unsized object
> Can you confirm you see the same?

No, I cannot confirm that.  

If you see a problem that you want us to act on, post real code that we can run and reproduce the problem.

**Comment by hanak** [comment]
@newville Thank you for your response and patience.

> Sorry, that is not enough to go on.

I understand and I am sorry for this. Sadly, I do not have permission to share the actual code. Thus, I tried to recreate the behavior using standard EPICS waveform. And, this did not work based on your response.

> Is there such a thing as a waveform of strings?

Technically, the documentation (https://epics.anl.gov/base/R7-0/6-docs/waveformRecord.html) states that field type of value (field FTVL) can accept `STRING`. And, EPICS tool caget will read the PV and display its content as expected. And, pyepics works with it right except for that single case I run into (with numpy enabled).

> No, I cannot confirm that.

Thank you for trying the test I provided. It is rather mystery to me why it is happening. I just wish I could see what arrives at the line 616 (https://github.com/pyepics/pyepics/blob/master/epics/pv.py) in the variable `val` at your system when the waveform PV contains `["ITEM 1"]`, FTVL=STRING, NELM=128, NORD=1. Would it be possible to share, please? In my case, `val` is `"ITEM 1"`, type `str`.

> If you see a problem that you want us to act on, post real code that we can run and reproduce the problem.

I agree. Until I am able to provide you an example which allows you to reproduce the behavior, there is no case. Just before you close this issue, would it be possible to run the example and share the value of `val` (incl. type) you see on your side when the waveform PV VAL=`["ITEM 1"]`, FTVL=STRING, NELM=128, NORD=1, please?

**Comment by newville** [comment]
@hanak 

> Thank you for trying the test I provided

Um, what test?  I don't see that. You will have to post code that others can run.  

> Thank you for trying the test I provided. It is rather mystery to me why it is happening. I just wish I could see what 
> arrives at the line 616 (https://github.com/pyepics/pyepics/blob/master/epics/pv.py) in the variable val at your system 
> when the waveform PV contains ["ITEM 1"], FTVL=STRING, NELM=128, NORD=1. Would it be possible to share, please?
> In my case, val is "ITEM 1", type str.

Share what?  What do you want me to do?  Seriously: post actual code that can be run to show the problem.

I can completely believe that we don't currently support a waveform of strings.  To support that, we would need to test examples.    We have no examples, so we cannot support it.

**Comment by hanak** [comment]
@newville I thought I provided the example earlier (comment no. 3, branch https://github.com/hanak/pyepics/tree/waveform-string-error, file https://github.com/hanak/pyepics/blob/waveform-string-error/tests/test_pv_unittests.py, lines modified 144 - 155). Yet, the wording might have been unclear. I wish I have noticed that earlier. In order to avoid my sleep deprived brain to confuse any further, let me copy-paste the modified test here:
```
def test_get_string_waveform():
    write('String Array: \n')
    with no_simulator_updates():
        nelm = int(caget(f"{pvnames.string_arr_pv}.NELM"))

        pv = get_pv(pvnames.string_arr_pv)
        for length in [1, 2, 3, nelm]:
            expected = [f"ITEM {index}" for index in range(length)]
            pv.put(expected)
            time.sleep(0.1)
            val = pv.get()
            assert val.tolist() == expected, f"Length {length}"
```

May I ask you to run the test and see what arrives at the line 616 in https://github.com/pyepics/pyepics/blob/master/epics/pv.py, please?
* The test requires "softIoc st.cmd" as other tests do.
* The command `print(f"VAL [{val}] (type {type(val)})")` placed before the line 616 should provide the output I would like to get (even through PyTest might will attempt to hide stdout in case the test succeeds).

Further details (just in case they are needed):
* The simulator (`tests/Setup/simulator.py`) was not running in background while running the test. 
* I used Python 3.10.11
* I used Python packages: numpy 2.0.0, pytest 8.2.2
* I used pyepics branch: master, commit eace0a176312965bfc5c1f0e7cd889397a46dc07 (tweak motor button sizes for linux)
* The test was executed on Windows 11

**Comment by newville** [comment]
@hanak will be fixed with #270.

**Comment by hanak** [comment]
@newville Thank you. It looks good. I have added a few minor comments for consideration.
